core/arch/arm/kernel/generic_entry_a32.S

```diff
@@ -218,6 +193,63 @@ END_FUNC reset
 		add	sp, r1, r2
 	.endm
 
+	/*
+	 * Cache maintenance during entry: handle outer cache.
+	 * End address is exclusive: first byte not to be changed.
+	 * Note however arm_clX_inv/cleanbyva operate on full cache lines.
+	 */
+	.macro inval_cache_vrange vbase, vend
+#ifdef CFG_PL310
+		/* expect mmu if off (and va=pa) */
+		bl	cpu_mmu_enabled
+		cmp	r0, #0
+		ldrne	r0, =__LINE__
+		blne	panic_at_line
+
+		bl	pl310_base
+		ldr	r1, =\vbase
+		ldr	r2, =(\vend - 1)
+		bl	arm_cl2_invbypa
+#endif
+		ldr	r0, =\vbase
+		ldr	r1, =(\vend - 1)
+		bl	arm_cl1_d_invbyva
+	.endm
+
+	.macro flush_cache_vrange vbase, vend
+#ifdef CFG_PL310
+		/* check va=pa for base address of target area */
+		bl	cpu_mmu_enabled
+		cmp	r0, #0
+		beq	1f
+		ldr	r0, =\vbase
+		bl	virt_to_phys
+		ldr	r1, =\vbase
+		cmp	r0, r1
+		ldrne	r0, =__LINE__
+		blne	panic_at_line
```

Comment:
i feel this is a bit ugly. I will shrink the 'va=pa' test into a single macro for both cache inval and flush.